runtime/near-vm-logic/src/logic.rs

```diff
@@ -2689,6 +2697,62 @@ impl<'a> VMLogic<'a> {
         let new_used_gas = self.gas_counter.used_gas();
         self.gas_counter.process_gas_limit(new_burn_gas, new_used_gas)
     }
+
+    /// VM independent setup before performing executable loading.
+    /// 
+    /// Does VM independent checks that happen after the instantiation of
+    /// VMLogic but before loading the executable. This includes pre-charging gas
+    /// costs for loading the executable, which depends on the size of the WASM code.
+    pub fn before_loading_executable(
+        &mut self,
+        method_name: &str,
+        current_protocol_version: u32,
+        wasm_code_bytes: usize,
+    ) -> std::result::Result<(), VMError> {
+        if method_name.is_empty() {
+            let error =
+                VMError::FunctionCallError(near_vm_errors::FunctionCallError::MethodResolveError(
+                    near_vm_errors::MethodResolveError::MethodEmptyName,
+                ));
+            return Err(error);
+        }
+        if checked_feature!(
+            "protocol_feature_fix_contract_loading_cost",
+            FixContractLoadingCost,
+            current_protocol_version
+        ) {
+            if self.add_contract_loading_fee(wasm_code_bytes as u64).is_err() {
+                let error =
+                    VMError::FunctionCallError(near_vm_errors::FunctionCallError::HostError(
+                        near_vm_errors::HostError::GasExceeded,
+                    ));
+                return Err(error);
+            }
+        }
+        Ok(())
+    }
+
+    /// Legacy code to preserve old gas charging behaviour in old protocol versions.
+    pub fn after_loading_executable(
+        &mut self,
+        current_protocol_version: u32,
+        wasm_code_bytes: usize,
+    ) -> std::result::Result<(), VMError> {
+        if !checked_feature!(
+            "protocol_feature_fix_contract_loading_cost",
+            FixContractLoadingCost,
+            current_protocol_version
+        ) {
+            if self.add_contract_loading_fee(wasm_code_bytes as u64).is_err() {
```

Comment:
We are not actually looking at what the error returned was and always assuming that it is `GasExceeded`.  I wonder if we can do something better here.  Although this code is probably not part of this PR and probably already existed so does not need to fixed in this PR.

Reply:
Yeah I see what you mean.

I think passing on the error from inside would look cleaner on the surface. But we would have to consider carefully if this is an observable change. From a quick glance, it is not obvious to me that it could not also be a `HostError::IntegerOverflow`.

And the second consideration would be, if we actually want to return more specific host errors, or if we specifically want to reduce it to just "gas exceeded"...

So yeah, not something I want to tie together with this PR.

Reply:
I raised a GH issue about it.  I'll resolve the conversation.